### PR TITLE
[v0.26.0rc][BugFix] Validate stop and allowed token ids

### PR DESCRIPTION
**Pull request by QwertyJack**
## What this PR does / why we need it?

Backport of #15339 (main) and #16161 (`releases/v0.27.1rc`) to `releases/v0.26.0rc`.

Validate client-supplied `stop_token_ids` and `allowed_token_ids` against the model vocabulary before device-side logits indexing, so an out-of-vocab id can no longer trap the CANN `IndexCheck` kernel and kill the engine (related issue: #15200, upstream fix: vllm#54196).

The vLLM commit pinned by this branch predates the upstream fix, so the patch stays active here:

- pin of `releases/v0.26.0rc` (`.github/vllm-main-verified.commit`): d02df748bf9efd99022f1a062597dc3cb3808485
- upstream fix vllm#54196 (merge commit 5b0e5b69ac), not contained in the pin above

The patch self-disables once the bundled vLLM ships `_validate_stop_token_ids`, so it becomes a no-op when the pin advances.

## Does this PR introduce any user-facing change?

Invalid token ids are rejected during request validation (HTTP 400) instead of crashing the engine on the device side.

## How was this patch tested?

Cherry-picked the #15339 commits onto `releases/v0.26.0rc`, then ran the new unit test and the whole patch suite against the installed vLLM, whose `vllm/sampling_params.py` is byte-identical to the v0.26 pin `d02df748bf`:

```
$ PYTHONPATH=<vllm-ascend v0.26 worktree> python -m pytest tests/ut/patch/platform/test_patch_stop_token_ids_validation.py -q
21 passed

$ PYTHONPATH=<vllm-ascend v0.26 worktree> python -m pytest tests/ut/patch -q
302 passed, 14 skipped
```

`ruff check` and `ruff format --check` pass on the changed files. No hardware run was done; this is validation-layer code only.

- vLLM main: https://github.com/vllm-project/vllm/commit/d02df748bf9efd99022f1a062597dc3cb3808485

### CODE REVIEW

**Comment by QwertyJack** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: still in progress, retry /rerun after completion

**Review by kunpengW-code — APPROVED** [review]

File: vllm_ascend/patch/platform/patch_stop_token_ids_validation.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    vocab_size = model_config.get_vocab_size()
+    invalid_token_ids = [token_id for token_id in self.stop_token_ids if token_id < 0 or token_id >= vocab_size]
+
+    if invalid_token_ids:
+        raise VLLMValidationError(
+            f"token_id(s) {invalid_token_ids} in stop_token_ids contain "
+            f"out-of-vocab token ids. Vocabulary size: {vocab_size}",
+            parameter="stop_token_ids",
+            value=invalid_token_ids,
+        )
```
[the remark this answers is not in the document]
Comment (kunpengW-code) [reply vllm_ascend/patch/platform/patch_stop_token_ids_validation.py:52]: Not an issue, please ignore.

File: vllm_ascend/patch/platform/patch_stop_token_ids_validation.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    allowed_token_ids = self.allowed_token_ids
+    if allowed_token_ids is None:
+        return
+
+    if len(allowed_token_ids) == 0:
+        raise VLLMValidationError(
+            "allowed_token_ids is not None and empty!",
+            parameter="allowed_token_ids",
+            value=allowed_token_ids,
+        )
```
[the remark this answers is not in the document]
Comment (kunpengW-code) [reply vllm_ascend/patch/platform/patch_stop_token_ids_validation.py:65]: Not an issue, please ignore.

File: vllm_ascend/patch/platform/patch_stop_token_ids_validation.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    vocab_size = model_config.get_vocab_size()
+    invalid_token_ids = [token_id for token_id in allowed_token_ids if token_id < 0 or token_id >= vocab_size]
+    if invalid_token_ids:
+        raise VLLMValidationError(
+            "allowed_token_ids contains out-of-vocab token id!",
+            parameter="allowed_token_ids",
+            value=invalid_token_ids,
+        )
```
[the remark this answers is not in the document]
Comment (kunpengW-code) [reply vllm_ascend/patch/platform/patch_stop_token_ids_validation.py:84]: Not an issue, please ignore.

File: vllm_ascend/patch/platform/patch_stop_token_ids_validation.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+if not hasattr(SamplingParams, "_validate_stop_token_ids"):
+    SamplingParams._validate_stop_token_ids = _validate_stop_token_ids
+    SamplingParams._validate_allowed_token_ids = _validate_allowed_token_ids
+
+    _orig_verify = SamplingParams.verify
+
+    def _verify(self, model_config, *args, **kwargs):
+        _validate_stop_token_ids(self, model_config)
+        _validate_allowed_token_ids(self, model_config)
+        return _orig_verify(self, model_config, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (kunpengW-code) [reply vllm_ascend/patch/platform/patch_stop_token_ids_validation.py:100]: Not an issue, please ignore.
